DSpace/dspace-angular
# Need alternative link index to make use of embeds

##  References

This PR closes #626

##  Description

This PR adds an extra index for alternative links to objects stored in the object cache.

##  Instructions for Reviewers

### Changes
Some objects can be requested by different links than just their self link. For example, in the `_links` section of an item, you will find a link to its owning collection formulated as the self link of the item, appended by `/owningCollection`. 

```
"_links": {
  "owningCollection": {
    "href": "https://example-server.org/api/core/items/0136fe00-43ca-439e-ae6c-b2b08c2b8f5e/owningCollection"
  },
  "self": {
    "href": "https://example-server.org/api/core/items/0136fe00-43ca-439e-ae6c-b2b08c2b8f5e"
  }
}  
```
When retrieving this collection, you'll notice the self link is not the same as the link mentioned in item's link section:
```
"_links": {
    "self": {
      "href": "https://example-server.org/api/core/collections/282164f5-d325-4740-8dd1-fa4d6d3e7200"
    }
  }
  ```
  
The owningCollection link mentioned in `_links` section of the item is an example of an alternative link. 

Previously, when for example, the owningCollection was embedded inside the item REST response, the embedded collection was stored in the object cache using its self link. Subsequently, the frontend would request this collection using its alternative link because the alternative link was not known to the cache. 

This PR adds an index to map the alternative links on the matching self links. For the example mentioned above, the entry of this index would look something like this:

| https://example-server.org/api/core/items/0136fe00-43ca-439e-ae6c-b2b08c2b8f5e/owningCollection | https://example-server.org/api/core/collections/282164f5-d325-4740-8dd1-fa4d6d3e7200 |
|-------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------|

To keep this index up to date, the following changes were made to the ObjectCacheService: 

* When something is added to the object cache, its alternative link is compared to its self link. If they differ:
	* A new entry is added to the index. 
	* The alternative link is also added to a list in the ObjectCacheEntry's state.

* When something is removed from the object cache: 
	* All links in the _links section of the object are removed from the new index, if they existed
	* The array of alternative links in the ObjectCacheEntry are also removed from the index

 ### How to test
A simple way of testing this PR is by visiting an item page, where the owningCollection is always embedded and referenced by an alternativeLink. 
If you go to a previous version of the code base, you'll notice a request is sent out to the alternative link ([itemLink]/owningCollection). This request has become redundant in this PR, and will not show up in the network tab of your browser anymore. If you look at the store, you will also be able to see the mapping in `core.index.object/alt-link-to-self-link`.

To test if the index is actually kept up to date, you could remove a relationship from the item on its edit page. This causes the item to be decached, which should also remove the alternative link to its owningCollection from the index in the store.

### Lists
Unfortunately, this PR does not solve the same issue for alternative links to object lists. Lists are not directly stored in the object cache, and can therefore not be mapped in the index.

##  Checklist

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests). Exceptions may be made if previously agreed upon.
